streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -744,18 +809,58 @@ private void assignTasksToClients(final Set<String> allSourceTopics,
                 }
                 state.addPreviousActiveTasks(previousActiveTasks);
             }
-        }
 
-        log.debug("Assigning tasks {} to clients {} with number of replicas {}",
-            partitionsForTask.keySet(), states, assignmentConfigs.numStandbyReplicas);
+            state.computeTaskLags(allTaskEndOffsetSums);
+            states.put(uuid, state);
+        }
+        return states;
+    }
 
-        // assign tasks to clients
-        final StickyTaskAssignor<UUID> taskAssignor =
-            new StickyTaskAssignor<>(states, partitionsForTask.keySet(), standbyTaskIds);
-        taskAssignor.assign(assignmentConfigs.numStandbyReplicas);
+    private Map<TaskId, Long> computeEndOffsetSumsByTask(final Map<TopicPartition, ListOffsetsResultInfo> endOffsets,
+                                                         final Map<TaskId, Set<TopicPartition>> changelogsByStatefulTask) {
+        final Map<TaskId, Long> taskEndOffsetSums = new HashMap<>();
+        for (final Map.Entry<TaskId, Set<TopicPartition>> taskEntry : changelogsByStatefulTask.entrySet()) {
+            final TaskId task = taskEntry.getKey();
+            taskEndOffsetSums.put(task, 0L);
+
+            for (final TopicPartition changelog : taskEntry.getValue()) {
+                final ListOffsetsResultInfo offsetResult = endOffsets.get(changelog);
+                if (offsetResult == null) {
+                    log.warn("Failed to fetch end offsets for changelog partition {}", changelog);
+                    continue;
+                }
```

Comment:
Ah, I was wondering whether `listOffsets` was all-or-none but forgot to look it up. In that case, I agree this should be an `IllegalStateException`.
On a related note, I've been wondering how (or if) we should handle the case where `listOffsets` fails. I think it will make the app too brittle if we throw an exception on failure. Some alternatives I can think of:
1) Just retry until we succeed, or hit some limit (either on number of retries or total time)
2) If the first try fails, just give up and fall back on the  current sticky algorithm
3) If the first try fails, just give everyone back their previous assignment and try again on the next rebalance
4) Some combination of the above
I think option 3) sounds best, although we could retry a few times before giving up on that rebalance. Thoughts?

Reply:
I agree option 3 looks the best, and also agree it would be good to have some retry, although perhaps the inbuilt AdminClient retries are sufficient for this case.

Reply:
That's determined by the `retries` config? That seems reasonable (although it defaults to 0? Huh)